### PR TITLE
Rework and upgrade GCP deployment

### PR DESCRIPTION
**Pull request by alicethorne-ab**
The goals of this PR are to:
* Upgrade the marketplace tools and helm charts to the latest versions
* Allow easier upgrade paths on GCP
* Overall code cleanup of the files

Issues I've run into during this work (and workarounds):
* The `deployer` image is immutable, but also _does not expire_, which makes redeploying/upgrading the SCIM Bridge through the GCP Marketplace difficult. Long story short, upgrading-in-place fails currently due to the `deployer` pod sticking around long after it's completed it's job. Kubernetes has no widely-available built-in options to clear `Complete` jobs like `deployer`s (there is technically an "alpha" feature available, but it's not enabled on GKE). I thus went with another workaround, which was to create a `CronJob` which deletes the `deployer` after a few minutes, allowing a fresh one to be deployed at-will. _This allows the deployment to be upgraded in-place on GCP!_ 🎉 
* The `marketplace-tools` vendor submodule no longer contains the code we need, and forking off of `marketplace-tools-example` isn't ideal because it contains a lot of code and things we don't need, which would lead to issues with merging things in. Unfortunately, we'll just need to manually copy over the files when upgrading. The GCP Marketplace documentation is not terribly clear on what to do in these situations, and they seem to want you to copy these files over anyways. Thus, the `vendor/` directory has been removed.

There are still some rough edges on this PR, but I wanted to put this up to show current progress.

### CODE REVIEW

File: op-scim-bridge/chart/op-scim-bridge/templates/op-scim-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: op-scim-bridge
```
Comment (pastjean) [inline op-scim-bridge/chart/op-scim-bridge/templates/op-scim-deployment.yaml:4]: There should be a variable here in the name, in case someone wants to deploy 2 bridges in the same cluster

File: op-scim-bridge/chart/op-scim-bridge/templates/redis-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: op-scim-redis
```
Comment (pastjean) [inline op-scim-bridge/chart/op-scim-bridge/templates/redis-deployment.yaml:4]: Same as above 

```suggestion
  name: {{ .Values.name }}-op-scim-redis
```

File: op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+apiVersion: batch/v1beta1
+kind: CronJob
+metadata:
+  name: deployer-cleanup # deletes stale deployers to improve upgradeability of the SCIM bridge
```
Comment (pastjean) [inline op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml:4]: Don't know if we want multiple of this one 

```suggestion
  name: deployer-cleanup # deletes stale deployers to improve upgradeability of the SCIM bridge
```

File: op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+spec:
+  schedule: "*/1 * * * *" # once a minute
+  jobTemplate:
+    spec:
+      template:
+        spec:
+          containers:
+          - name: kubectl-runner
+            image: bitnami/kubectl
+            command: ["sh", "-c", "kubectl delete jobs $(kubectl get jobs | awk '$2 ~ 1/1' | awk '{print $1}')"] # removes "Complete" deployer pods
```
Comment (pastjean) [inline op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml:14]: I'm missing the arguments here, we delete everything ?

I feel this might just delete all jobs

**Review by pastjean — DISMISSED** [review]

File: app.Makefile [hunk 5]
```diff
@@ -0,0 +1,131 @@
+ifndef __APP_MAKEFILE__
```
Comment (grellyd) [inline app.Makefile:1]: I see all these Makefiles are from the GCP project, and were previously included directly from the vendored directory. 

Do we need all of them? I see only three directly in use in the makefile. Do some depend on the others?

Additionally can we take them out of the root and put them in a subfolder? That would make their difference compared to the actual makefile a bit clearer.

Finally how can we update these files? What is the procedure? What version are these from? Is that what the `MARKETPLACE_TOOLS_TAG` is for? Can we turn that procedure into a simple Makefile command? 

In any case, please document how you have upgraded the marketplace tools so it may be repeated easily.

File: op-scim-bridge/chart/op-scim-bridge/Chart.yaml [hunk 6]
```diff
@@ -1,5 +1,6 @@
-apiVersion: v1
```
Comment (grellyd) [inline op-scim-bridge/chart/op-scim-bridge/Chart.yaml:1]: Did you have any luck leveraging our existing Helm Chart into this repo? 

See: https://gitlab.1password.io/dev/b5/op/-/issues/866

File: op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+apiVersion: batch/v1beta1
```
Comment (grellyd) [inline op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml:1]: Awesome. Thanks for making this exist!

File: op-scim-bridge/chart/op-scim-bridge/templates/op-scim-deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        command: ["/op-scim/op-scim"]
+        env:
+        - name: "OP_PORT"
+          value: "8080"
+        - name: "OP_SESSION"
+          value: "{{ .Values.scimUserHome }}/{{ .Values.scimsessionFile }}"
+        - name: "OP_REDIS_HOST"
+          value: "{{ .Values.name }}-op-scim-redis-svc"
+        - name: "OP_ONE_CLICK"
+          value: "true"
```
Comment (grellyd) [inline op-scim-bridge/chart/op-scim-bridge/templates/op-scim-deployment.yaml:39]: Same question as above: why do we no longer set the `OP_DOMAIN` value from `OP_ACCOUNT_DOMAIN`?

In the old version it was:

```
 - name: OP_DOMAIN
          value: {{ .Values.OP_ACCOUNT_DOMAIN }}
 ```

**Review by grellyd — COMMENTED** [review]
Looking very promising. 

Couple of questions! Some are a bit rambly, as I currently lack the context of some of the upgrade choices. Let's hop on a call if anything is unclear.

File: README.md [hunk 9]
```diff
@@ -75,66 +78,37 @@ Tags are updated regularly and may vary. The `latest` tag is not always updated,
...
+It is now possible to use `mpdev` to install our application on your Kubernetes cluster on GCP. This allows you to test our application outside of the marketplace.
 
-As defined in [schema.yaml](./op-scim-bridge/schema.yaml) our application requires three arguments: `APP_INSTANCE_NAME`, `NAMESPACE`, and `OP_ACCOUNT_DOMAIN`. These parallel the values chosen in the user interface when deploying from the marketplace. This example uses `opscim.b5test.com`, but please replace that with your testing account, as otherwise your bridge will not allow the authentication.
+As defined in [schema.yaml](./op-scim-bridge/schema.yaml) our application requires two arguments: `name` and `namespace`. These parallel the values chosen in the user interface when deploying from the marketplace. 
 
-```
-mpdev install --deployer=gcr.io/op-scim-bridge/op-scim-bridge/deployer:latest --parameters='{"APP_INSTANCE_NAME": "mpdev", "NAMESPACE": "default", "OP_ACCOUNT_DOMAIN": "opscim.b5test.com"}'
+```bash
+mpdev install --deployer=gcr.io/op-scim-bridge/op-scim-bridge/deployer:latest --parameters='{"name": "mpdev", "namespace": "default"}'
```
[the remark this answers is not in the document]
Comment (alicethorne-ab) [reply README.md]: Re-added!

File: op-scim-bridge/chart/op-scim-bridge/Chart.yaml (hunk 6 above)
Comment (alicethorne-ab) [reply op-scim-bridge/chart/op-scim-bridge/Chart.yaml:1]: Like you posted there, the `do-block-storage` option will need to be worked around in terms of the DO/GCP split. We can likely override it in some way when packaging it, but that'll have to be a larger conversation.

File: app.Makefile (hunk 5 above)
Comment (alicethorne-ab) [reply app.Makefile]: Added a readme, moved files to `tools/`, and removed files that weren't needed!

File: op-scim-bridge/schema.yaml [hunk 10]
```diff
@@ -1,32 +1,47 @@
...
-    default: $REGISTRY:$TAG
-    x-google-marketplace:
-      type: IMAGE
-  IMAGE_REDIS:
-    type: string
-    default: $REGISTRY/redis:$TAG
-    x-google-marketplace:
-      type: IMAGE
-  OP_ACCOUNT_DOMAIN:
+  accountdomain:
```
Comment (grellyd) [inline op-scim-bridge/schema.yaml:36]: What are your thoughts on camel casing these variables for readability? Do you know of any kubernetes values naming standards?

Curious to hear more re the Property naming changes.

File: tools/README.md [hunk 11]
```diff
@@ -0,0 +1,20 @@
+# Updating these tools
```
Comment (grellyd) [inline tools/README.md:1]: Yay thanks for adding this!

**Review by grellyd — COMMENTED** [review]
Looking great! One quick question on naming. 

The next step is to test this via MPDev myself. I'll circle back to that later today.

File: op-scim-bridge/schema.yaml (hunk 10 above)
Comment (alicethorne-ab) [reply op-scim-bridge/schema.yaml]: The overall idea was to bring the variable names more in-line with what is presented in Google's Marketplace examples, as seen here: https://github.com/GoogleCloudPlatform/marketplace-k8s-app-example/blob/master/wordpress/schema.yaml

You're correct, however, in that `accountdomain` should be camelcased. I've gone ahead and done that in the latest commit.

File: README.md [hunk 12]
```diff
@@ -75,66 +78,37 @@ Tags are updated regularly and may vary. The `latest` tag is not always updated,
...
 You can validate your installation by running `mpdev doctor`
 
-### mpdev install
+## Install
 
-It is now possible to use mpdev to install our application on your Kubernetes cluster on GCP. This allows you to test our application outside of the marketplace.
+It is now possible to use `mpdev` to install our application on your Kubernetes cluster on GCP. This allows you to test our application outside of the marketplace.
 
-As defined in [schema.yaml](./op-scim-bridge/schema.yaml) our application requires three arguments: `APP_INSTANCE_NAME`, `NAMESPACE`, and `OP_ACCOUNT_DOMAIN`. These parallel the values chosen in the user interface when deploying from the marketplace. This example uses `opscim.b5test.com`, but please replace that with your testing account, as otherwise your bridge will not allow the authentication.
+As defined in [schema.yaml](./op-scim-bridge/schema.yaml) our application requires three arguments: `name`, `namespace`, and `accountdomain`. These parallel the values chosen in the user interface when deploying from the Marketplace.
```
Comment (grellyd) [inline README.md:93]:
```suggestion
As defined in [schema.yaml](./op-scim-bridge/schema.yaml) our application requires three arguments: `name`, `namespace`, and `accountDomain`. These parallel the values chosen in the user interface when deploying from the Marketplace. 
```

File: README.md [hunk 13]
```diff
@@ -75,66 +78,37 @@ Tags are updated regularly and may vary. The `latest` tag is not always updated,
...
+It is now possible to use `mpdev` to install our application on your Kubernetes cluster on GCP. This allows you to test our application outside of the marketplace.
 
-As defined in [schema.yaml](./op-scim-bridge/schema.yaml) our application requires three arguments: `APP_INSTANCE_NAME`, `NAMESPACE`, and `OP_ACCOUNT_DOMAIN`. These parallel the values chosen in the user interface when deploying from the marketplace. This example uses `opscim.b5test.com`, but please replace that with your testing account, as otherwise your bridge will not allow the authentication.
+As defined in [schema.yaml](./op-scim-bridge/schema.yaml) our application requires three arguments: `name`, `namespace`, and `accountdomain`. These parallel the values chosen in the user interface when deploying from the Marketplace. 
 
-```
-mpdev install --deployer=gcr.io/op-scim-bridge/op-scim-bridge/deployer:latest --parameters='{"APP_INSTANCE_NAME": "mpdev", "NAMESPACE": "default", "OP_ACCOUNT_DOMAIN": "opscim.b5test.com"}'
+```bash
+mpdev install --deployer=gcr.io/op-scim-bridge/op-scim-bridge/deployer:latest --parameters='{"name": "mpdev", "namespace": "default", "accountdomain": "testing.1password.com" }'
```
Comment (grellyd) [inline README.md:96]:
```suggestion
mpdev install --deployer=gcr.io/op-scim-bridge/op-scim-bridge/deployer:latest --parameters='{"name": "mpdev", "namespace": "default", "accountDomain": "testing.1password.com" }'
```

**Review by grellyd — COMMENTED** [review]
Just starting to manually test this deployer now. 

One other issue to consider: we don't generally push the `latest` tag for the deployer. The image that is there currently is from Nov 30, 2018. When actually using these instructions I always swap `deployer:latest` for the tag I am deploying `deployer:1.6.2`

**Comment by grellyd** [comment]
Screenshot 2021-02-24 at 1 34 59 PM

The Cronjob seems to be re-running after completion and is erroring out.

I tested this via:

`REGISTRY=gcr.io/op-scim-bridge PUBLIC_TAG=1.6.2 TAG=gcp-mr-25 make app/build`

then installed that via:

`mpdev install --deployer=gcr.io/op-scim-bridge/op-scim-bridge/deployer:gcp-mr-25 --parameters='{"name": "mpdev", "namespace": "default", "accountDomain": "opscim-clone.b5test.com" }'` 

You should be able to reproduce this failure with my deployer tag `gcp-mr-25`. I deployed it on a pre-existing cluster after removing all the resources via:

`kubectl delete replicaset --all && kubectl delete jobs --all && kubectl delete deployment --all && kubectl delete service --all && kubectl delete pod --all && kubectl delete pvc --all`

**Comment by alicethorne-ab** [comment]
I made the cronjob a little more robust in terms of checking that there's actually something to delete before it runs, as well as limiting it to the namespace of the instance, and limiting it to only deleting jobs with "deployer" in the title. Should work a bit better now.

**Comment by grellyd** [comment]
The Cronjob seems to be re-running after completion and is erroring out.
Reposting for review clarity: 

It seems there is an issue with the CronJob.

cronFailure

I tested this via:

REGISTRY=registry PUBLIC_TAG=1.6.2 TAG=gcp-mr-25 make app/build

then installed that via:

mpdev install --deployer=registry:gcp-mr-25 --parameters='{"name": "mpdev", "namespace": "default", "accountDomain": "subdomain.domain.com" }'

You should be able to reproduce this failure with my deployer tag gcp-mr-25. I deployed it on a pre-existing cluster after removing all the resources via:

kubectl delete replicaset --all && kubectl delete jobs --all && kubectl delete deployment --all && kubectl delete service --all && kubectl delete pod --all && kubectl delete pvc --all

File: op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml [hunk 14]
```diff
@@ -1,15 +1,22 @@
...
     spec:
       template:
+        metadata:
+          labels: *CronLabels
         spec:
           containers:
           - name: kubectl-runner
             image: bitnami/kubectl
-            command: ["bash", "-c", "if [[ `kubectl get pods -n {{ .Values.namespace }} | grep -i deployer | awk '$2 ~ 1/1' | awk '{print $1}'` ]]; then kubectl delete jobs -n {{ .Values.namespace }} $(kubectl get jobs -n {{ .Values.namespace }} | grep -i deployer | awk '$2 ~ 1/1' | awk '{print $1}'); fi"]
+            command: ["bash", "-c", "if [[ `kubectl get pods -n {{ .Values.namespace }} | awk '$1 ~ /{{ .Values.name }}-deployer/ && $2 ~ /0.*1/ { print $1 }'` ]]; then kubectl delete jobs -n {{ .Values.namespace }} $(kubectl get jobs -n {{ .Values.namespace }} | awk '$1 ~ /{{ .Values.name }}-deployer/ && $2 ~ /0.*1/ {print $1}' 2>/dev/null) 2>/dev/null; fi; exit 0"]
```
Comment (grellyd) [inline op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml:21]: The diff here is a bit hard to review; it's just commands piped together. Is there a cleaner way to go about doing this? EG: where the changes will be presented in a nice structured file rather than a single line.

File: op-scim-bridge/Makefile [hunk 15]
```diff
@@ -12,56 +12,55 @@ APP_GCS_PATH ?= $(GCS_URL)/$(APP_ID)/$(TAG)
...
+						 .build/var/TAG \
+						 apptest/deployer/* \
+						 deployer/* \
+						 schema.yaml \
+						 | .build/op-scim
+	$(call print_target, $@)
 	docker build \
-	    --build-arg REGISTRY="$(REGISTRY)/op-scim-bridge" \
-	    --build-arg TAG="$(TAG)" \
-	    --build-arg PUBLIC_TAG="$(PUBLIC_TAG)" \
```
Comment (grellyd) [inline op-scim-bridge/Makefile]: Can you explain why we are removing the `PUBLIC_TAG`? Do we not need it when building the deployer?

The usage in the past has been `PUBLIC_TAG=$version_I_want_to_use $TAG=$my_development_tag_to_tag`

Correct me if I'm wrong, the current setup will only allow us to tag production releases of the deployer, not development releases. 

I note you still use the $PUBLIC_TAG below in the `build/op-scim/op-scim-bridge` command.

What am I missing?

**Review by grellyd — COMMENTED** [review]
Some more questions for you! I'll give this a test later today.

File: op-scim-bridge/Makefile (hunk 15 above)
Comment (alicethorne-ab) [reply op-scim-bridge/Makefile]: The `deployer` image actually uses the `deployer/Dockerfile` to be built, and one thing that I noticed is that the `Dockerfile` does not actually ingest the `PUBLIC_TAG` environment variable at any point during the image being built. This means that `PUBLIC_TAG` is simply not used during that command, so I removed it to be more consistent. I actually did not make many modifications to the `Dockerfile`, and its lack of `PUBLIC_TAG` isn't something I decided on. We could talk about adding it in, but that's simply the way it has been until now.

File: op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml (hunk 14 above)
Comment (alicethorne-ab) [reply op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml:21]: I could break it up into multiple lines. Would that work?

Comment (grellyd) [reply op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml:21]: That would probably be a bit clearer. However at the end of the day, I'll leave the format up to you. Whatever you think is the most clear and best outcome.

I just wanted to note that this is somewhat hard to parse as a viewer, and that if we are doing such a complex task in a `command` block, that smells to me as if we are not doing something right. However I don't know what the 'right' way to do this is either. 

I'll do a little looking around to see if there is a better way to do this.

**Review by grellyd — COMMENTED** [review]
Halfway through testing. Looking good!

File: op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml (hunk 14 above)
Comment (alicethorne-ab) [reply op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml:21]: I looked high and low to see if there was a better solution. Most solutions seem to rely on either the software auto-updating itself (through various advanced mechanisms), or in features that never quite made the cut to mainline K8s. 

The command here is relatively simple: get the `deployer` and delete it. It could be reduced further to a single line. However, I'm erring on the side of caution and _only_ attempting to when the deployer is _completed_, which requires additional checks. The issue is that we don't want it deleting a `deployer` when it's still deploying.

Overall there are quite a few things with the GCP Marketplace Tools in general which have room for improvement, but are largely out of our hands. Reducing this "delete deployer" check to a single command was about as simple as I could make it.

**Comment by grellyd** [comment]
Screenshot 2021-03-08 at 11 08 36 AM

Looking good!

You said that this:
> allows the deployment to be upgraded in-place on GCP! 

How can I test that? What is the procedure?

File: op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml (hunk 14 above)
Comment (grellyd) [reply op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml:21]: I appreciate your hard work on this topic! Thank you for taking such a detailed look. I didn't come up with anything better either.

Comment (alicethorne-ab) [reply op-scim-bridge/chart/op-scim-bridge/templates/deployer-cleanup.yaml]: I went ahead and moved the "script" to an actual `Config` which makes it a lot more readable. We use a similar mechanism in the verifier scripts already (in `apptest`), so I largely took from that example.

**Comment by grellyd** [comment]
As we chatted on the call, while using the default namespace I'm getting a permissions error for the default service account. As I understand it, the solution is to define a service account for the job. 
```
>kubectl logs job.batch/mpdev-cleanup-1615837020 -f
Error from server (Forbidden): jobs.batch is forbidden: User "system:serviceaccount:default:default" cannot list resource "jobs" in API group "batch" in the namespace "default"
checking for deployer
deployer not detected, skipping
```

**Review by grellyd — DISMISSED** [review]
LGTM! ✅ 

We have tested this many times over 😂

**Review by devillexio — DISMISSED** [review]

**Comment by grellyd** [comment]
It was an oversight on my part to fail to run `mpdev verify` before now. However any new version I create does not pass mpdev verify with the error:

```
NFO Wait 300.0 seconds for the application 'apptest-fcmrddh5' to get into ready state
    kubectl get "applications.app.k8s.io/apptest-fcmrddh5"
      --namespace="apptest-fcmrddh5"
      --output=json
Traceback (most recent call last):
  File "/bin/wait_for_ready.py", line 229, in <module>
    main()
  File "/bin/wait_for_ready.py", line 62, in main
    '''.format(kind, args.namespace, args.name)).json()
  File "/bin/bash_util.py", line 47, in __init__
    self._run()
  File "/bin/bash_util.py", line 58, in _run
    raise CommandException(self._exitcode, error_message)
bash_util.CommandException: Using /opt/kubectl/1.18/kubectl (server=1.18)
Error from server (Forbidden): persistentvolumes is forbidden: User "system:serviceaccount:apptest-fcmrddh5:apptest-fcmrddh5-deployer-sa" cannot list resource "persistentvolumes" in API group "" at the cluster scope
```

I am getting an success for `mpdev verify` on the `1.6.2` tag in development registry.

**Comment by grellyd** [comment]
Screenshot 2021-04-07 at 6 16 06 PM

I am also seeing some file conflicts.

**Comment by alicethorne-ab** [comment]
Those errors were fixed and a result of me forking the previous work on StatefulSets out of this branch. `mpdev verify` now works as expected.

**Comment by grellyd** [comment]
Screenshot 2021-04-08 at 12 02 31 PM

Works and verifies! 

LGTM ✅

**Review by grellyd — APPROVED** [review]
